File: src/pulso/loader/lookups.clj
(ns pulso.loader.lookups
  (:require [next.jdbc :as jdbc]))

(def ^:private source-cache (atom {}))
(def ^:private device-cache (atom {}))
(def ^:private record-type-cache (atom {}))
(def ^:private unit-cache (atom {}))

(defn reset-caches!
  "Clears all lookup caches."
  []
  (reset! source-cache {})
  (reset! device-cache {})
  (reset! record-type-cache {})
  (reset! unit-cache {}))

(defn ensure-source-id!
  "Returns the source ID for the given name+version, inserting if needed."
  [ds source-name source-version]
  (when source-name
    (let [cache-key [source-name source-version]]
      (or (get @source-cache cache-key)
          (let [row (jdbc/execute-one! ds
                      ["INSERT INTO source (name, version) VALUES (?, ?)
                        ON CONFLICT (name, version) DO UPDATE SET name = EXCLUDED.name
                        RETURNING id"
                       source-name source-version])]
            (swap! source-cache assoc cache-key (:source/id row))
            (:source/id row))))))

(defn ensure-device-id!
  "Returns the device ID for the given raw text, inserting if needed."
  [ds raw-text]
  (when raw-text
    (or (get @device-cache raw-text)
        (let [row (jdbc/execute-one! ds
                    ["INSERT INTO device (raw_text) VALUES (?)
                      ON CONFLICT (raw_text) DO UPDATE SET raw_text = EXCLUDED.raw_text
                      RETURNING id"
                     raw-text])]
          (swap! device-cache assoc raw-text (:device/id row))
          (:device/id row)))))

(defn ensure-record-type-id!
  "Returns the record_type ID for the given identifier, inserting if needed."
  [ds identifier]
  (when identifier
    (or (get @record-type-cache identifier)
        (let [row (jdbc/execute-one! ds
                    ["INSERT INTO record_type (identifier) VALUES (?)
                      ON CONFLICT (identifier) DO UPDATE SET identifier = EXCLUDED.identifier
                      RETURNING id"
                     identifier])]
          (swap! record-type-cache assoc identifier (:record_type/id row))
          (:record_type/id row)))))

(defn ensure-unit-id!
  "Returns the unit ID for the given name, inserting if needed."
  [ds unit-name]
  (when unit-name
    (or (get @unit-cache unit-name)
        (let [row (jdbc/execute-one! ds
                    ["INSERT INTO unit (name) VALUES (?)
                      ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name
                      RETURNING id"
                     unit-name])]
          (swap! unit-cache assoc unit-name (:unit/id row))
          (:unit/id row)))))

File: apps/etl-python/pulso/loader/lookups.py
from pulso import db

_source_cache = {}
_device_cache = {}
_record_type_cache = {}
_unit_cache = {}


def reset_caches():
    _source_cache.clear()
    _device_cache.clear()
    _record_type_cache.clear()
    _unit_cache.clear()


def _upsert_id(pool, sql, params, cache, cache_key):
    if cache_key in cache:
        return cache[cache_key]
    with db.get_conn(pool) as conn:
        with conn.cursor() as cur:
            cur.execute(sql, params)
            row_id = cur.fetchone()[0]
        conn.commit()
    cache[cache_key] = row_id
    return row_id


def ensure_source_id(pool, source_name, source_version):
    if source_name is None:
        return None
    cache_key = (source_name, source_version)
    return _upsert_id(
        pool,
        """INSERT INTO source (name, version) VALUES (%s, %s)
           ON CONFLICT (name, version) DO UPDATE SET name = EXCLUDED.name
           RETURNING id""",
        (source_name, source_version),
        _source_cache,
        cache_key,
    )


def ensure_device_id(pool, raw_text):
    if raw_text is None:
        return None
    return _upsert_id(
        pool,
        """INSERT INTO device (raw_text) VALUES (%s)
           ON CONFLICT (raw_text) DO UPDATE SET raw_text = EXCLUDED.raw_text
           RETURNING id""",
        (raw_text,),
        _device_cache,
        raw_text,
    )


def ensure_record_type_id(pool, identifier):
    if identifier is None:
        return None
    return _upsert_id(
        pool,
        """INSERT INTO record_type (identifier) VALUES (%s)
           ON CONFLICT (identifier) DO UPDATE SET identifier = EXCLUDED.identifier
           RETURNING id""",
        (identifier,),
        _record_type_cache,
        identifier,
    )


def ensure_unit_id(pool, unit_name):
    if unit_name is None:
        return None
    return _upsert_id(
        pool,
        """INSERT INTO unit (name) VALUES (%s)
           ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name
           RETURNING id""",
        (unit_name,),
        _unit_cache,
        unit_name,
    )
